controller/konnect/reconciler_generic.go

```diff
@@ -354,6 +360,150 @@ func updateStatusWithCondition[T interface {
+		return mo.Some(e.Spec.ControlPlaneRef)
+	default:
+		panic(fmt.Sprintf("unsupported entity type %T", e))
+	}
+}
+
+func handleControlPlaneRef[T SupportedKonnectEntityType, TEnt EntityType[T]](
+	ctx context.Context,
+	cl client.Client,
+	ent TEnt,
+) (ctrl.Result, error) {
+	cpRef, ok := getControlPlaneRef(ent).Get()
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	switch cpRef.Type {
+	case configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef:
+		cp := konnectv1alpha1.KonnectControlPlane{}
+		nn := types.NamespacedName{
+			Name:      cpRef.KonnectNamespacedRef.Name,
+			Namespace: cpRef.KonnectNamespacedRef.Namespace,
+		}
+		if nn.Namespace == "" {
+			nn.Namespace = ent.GetNamespace()
+		}
+		if err := cl.Get(ctx, nn, &cp); err != nil {
+			if res, errStatus := updateStatusWithCondition(
+				ctx, cl, ent,
+				ControlPlaneRefValidConditionType,
+				metav1.ConditionFalse,
+				ControlPlaneRefReasonInvalid,
+				err.Error(),
+			); errStatus != nil || res.Requeue {
+				return res, errStatus
+			}
+			return ctrl.Result{}, err
+		}
+
+		cond, ok := k8sutils.GetCondition(KonnectEntityProgrammedConditionType, &cp)
+		if !ok || cond.Status != metav1.ConditionTrue || cond.ObservedGeneration != cp.GetGeneration()  {
+			if res, errStatus := updateStatusWithCondition(
+				ctx, cl, ent,
+				ControlPlaneRefValidConditionType,
+				metav1.ConditionFalse,
+				ControlPlaneRefReasonInvalid,
+				fmt.Sprintf("Referenced ControlPlane %s is not programmed yet", nn),
+			); errStatus != nil || res.Requeue {
+				return res, errStatus
+			}
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		old := ent.DeepCopyObject().(TEnt)
+		if err := controllerutil.SetOwnerReference(&cp, ent, cl.Scheme(), controllerutil.WithBlockOwnerDeletion(true)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set owner reference: %w", err)
+		}
+		if err := cl.Patch(ctx, ent, client.MergeFrom(old)); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("failed to update status: %w", err)
+		}
+
+		type EntityWithControlPlaneRef interface {
+			SetControlPlaneID(string)
+			GetControlPlaneID() string
+		}
+		// TODO(pmalek): make this generic.
+		// CP ID is not stored in KonnectEntityStatus because not all entities
+		// have a ControlPlaneRef, hence the type constraints in the reconciler can't be used.
+		if resource, ok := any(ent).(EntityWithControlPlaneRef); ok {
+			resource.SetControlPlaneID(cp.Status.ID)
+		}
+
+		if res, errStatus := updateStatusWithCondition(
+			ctx, cl, ent,
+			ControlPlaneRefValidConditionType,
+			metav1.ConditionTrue,
+			ControlPlaneRefReasonValid,
+			fmt.Sprintf("Referenced ControlPlane %s is programmed", nn),
+		); errStatus != nil || res.Requeue {
+			return res, errStatus
+		}
+
+		return ctrl.Result{}, nil
+
+	default:
+		return ctrl.Result{}, fmt.Errorf("unimplemented ControlPlane ref type %q", cpRef.Type)
```

Comment:
This does make sense to me but we would be setting a precedent here. 

Please note that also not returning an error will not be accounted for in the reconciler metrics. So if a user relies on those instead of the logs, this error can not trigger an alert.

Reply:
Yeah, I see your point about the metrics. Let's leave it then.